### PR TITLE
Mtc 417 initial implementation

### PR DESCRIPTION
**Pull request by andrey-anokhin**

### CODE REVIEW

File: src/MarginTrading.Activities.Core/Settings/ActivitiesSettings.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        public ChaosSettings ChaosKitty { get; set; }
+        
+        [Optional]
+        public bool UseSerilog { get; set; }
+        
+        public CqrsSettings Cqrs { get; set; }
+        
+        public ConsumersSettings Consumers { get; set; }
+        
+        public SettingsServiceClient SettingsServiceClient { get; set; }
```
Comment (GSailen) [inline src/MarginTrading.Activities.Core/Settings/ActivitiesSettings.cs]: Use external section, which already exists.

File: src/MarginTrading.Activities.Core/Settings/SettingsServiceClient.cs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+using JetBrains.Annotations;
+using Lykke.SettingsReader.Attributes;
+
+namespace MarginTrading.Activities.Core.Settings
+{
+    [UsedImplicitly]
+    public class SettingsServiceClient
```
Comment (GSailen) [inline src/MarginTrading.Activities.Core/Settings/SettingsServiceClient.cs]: Let's introduce generic setting for any client here like ServiceClientSettings

File: src/MarginTrading.Activities.Producer/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,4 @@
+FROM microsoft/dotnet:2.1.5-aspnetcore-runtime
```
Comment (GSailen) [inline src/MarginTrading.Activities.Producer/Dockerfile]: 2.2.1 is the latest

File: src/MarginTrading.Activities.Producer/Infrastructure/CustomOperationIdOperationFilter.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿using Microsoft.AspNetCore.Mvc.Controllers;
+using Swashbuckle.AspNetCore.Swagger;
+using Swashbuckle.AspNetCore.SwaggerGen;
+
+namespace MarginTrading.Activities.Producer.Infrastructure
+{
+    public class CustomOperationIdOperationFilter : IOperationFilter
```
Comment (GSailen) [inline src/MarginTrading.Activities.Producer/Infrastructure/CustomOperationIdOperationFilter.cs]: [UsedImplicitly]

File: src/MarginTrading.Activities.Producer/Properties/AssemblyAttributes.cs [hunk 5]
```diff
@@ -0,0 +1,8 @@
+﻿using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("MarginTrading.OvernightSwapService.Tests")]
```
Comment (GSailen) [inline src/MarginTrading.Activities.Producer/Properties/AssemblyAttributes.cs]: Copy-paste )

File: src/MarginTrading.Activities.Services/Abstractions/IAssetPairsCacheService.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+using MarginTrading.SettingsService.Contracts.AssetPair;
+
+namespace MarginTrading.Activities.Services.Abstractions
+{
+    public interface IAssetPairsCacheService
+    {
+        //TODO: if contract will change, create internal domain model
+        AssetPairContract TryGetAssetPair(string assetPairId);
```
Comment (GSailen) [inline src/MarginTrading.Activities.Services/Abstractions/IAssetPairsCacheService.cs]: Please do not store contracts.. convert them to domain model, it's not very hard to do.

File: src/MarginTrading.Activities.Services/Abstractions/IDateService.cs [hunk 7]
```diff
@@ -0,0 +1,9 @@
+using System;
+
+namespace MarginTrading.Activities.Services.Abstractions
+{
+    public interface IDateService
```
Comment (GSailen) [inline src/MarginTrading.Activities.Services/Abstractions/IDateService.cs]: Both IDateService and ISystemClock are used - let's use one of them.

File: src/MarginTrading.Activities.Services/ActivitiesPublisher.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
+using System.Threading.Tasks;
+using Common;
+using MarginTrading.Activities.Core.Domain.Abstractions;
+
+namespace MarginTrading.Activities.Services
+{
+    public class ActivitiesPublisher : IMessageProducer<IActivity>
```
Comment (GSailen) [inline src/MarginTrading.Activities.Services/ActivitiesPublisher.cs]: Unused code

File: src/MarginTrading.Activities.Services/Extensions.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+            return value.HasValue 
+                ? value.Value.ToUiString(precision) 
+                : ((decimal) 0).ToUiString(precision);
+        }
+        
+        public static string ToUiString(this decimal value, int? precision)
+        {
+            precision = precision ?? DefaultDigitPrecision;
+            
+            var sharps = precision > 0
```
Comment (GSailen) [inline src/MarginTrading.Activities.Services/Extensions.cs]: Don't understand it. How could asset pair accuracy be less than zero?

File: src/MarginTrading.Activities.Core/Settings/CqrsContextNamesSettings.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+using Lykke.SettingsReader.Attributes;
+
+namespace MarginTrading.Activities.Core.Settings
+{
+    public class CqrsContextNamesSettings
+    {
+        [Optional] public string AccountsManagement { get; set; } = nameof(AccountsManagement);
+
+        [Optional] public string TradingEngine { get; set; } = nameof(TradingEngine);
```
Comment (GSailen) [inline src/MarginTrading.Activities.Core/Settings/CqrsContextNamesSettings.cs]: not used

File: src/MarginTrading.Activities.Services/Projections/OrdersProjection.cs [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+                _rabbitMqSubscriberService.GetJsonDeserializer<OrderHistoryEvent>());
+        }
+
+        private Task HandleOrderHistoryEvent(OrderHistoryEvent historyEvent)
+        {
+            var order = historyEvent.OrderSnapshot;
+
+            if (order == null)
+            {
+                _log.WriteWarning(
```
Comment (GSailen) [inline src/MarginTrading.Activities.Services/Projections/OrdersProjection.cs]: Can be written asynchronously.

File: src/MarginTrading.Activities.Services/Projections/PositionsProjection.cs [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+                _rabbitMqSubscriberService.GetJsonDeserializer<PositionHistoryEvent>());
+        }
+
+        private Task HandlePositionHistoryEvent(PositionHistoryEvent historyEvent)
+        {
+            var position = historyEvent.PositionSnapshot;
+
+            if (position == null)
+            {
+                _log.WriteWarning(
```
Comment (GSailen) [inline src/MarginTrading.Activities.Services/Projections/PositionsProjection.cs]: Write it async

File: src/MarginTrading.Activities.SqlRepositories/ActivitiesRepository.cs [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        {
+            _connectionString = connectionString;
+            _log = log;
+            
+            using (var conn = new SqlConnection(connectionString))
+            {
+                try { conn.CreateTableIfDoesntExists(CreateTableScript, TableName); }
+                catch (Exception ex)
+                {
+                    _log?.WriteErrorAsync(TableName, "CreateTableIfDoesntExists", null, ex);
```
Comment (GSailen) [inline src/MarginTrading.Activities.SqlRepositories/ActivitiesRepository.cs]: Use sync version here

File: src/MarginTrading.Activities.Broker/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,4 @@
+FROM microsoft/dotnet:2.1.5-aspnetcore-runtime
```
[the remark this answers is not in the document]
Comment (andrey-anokhin) [reply src/MarginTrading.Activities.Broker/Dockerfile]: Decided to leave as is

File: src/MarginTrading.Activities.Services/Abstractions/IAssetPairsCacheService.cs (hunk 6 above)
Comment (andrey-anokhin) [reply src/MarginTrading.Activities.Services/Abstractions/IAssetPairsCacheService.cs]: I think this is useless in this particular case. We do not expose these contracts to the client, it is not supposed to use this contract anywhere else. And if client library will change, we will need to fix mappings anyway.

File: src/MarginTrading.Activities.Services/Abstractions/IDateService.cs (hunk 7 above)
Comment (andrey-anokhin) [reply src/MarginTrading.Activities.Services/Abstractions/IDateService.cs]: ISystemClock was not used, only added to the container.

File: src/MarginTrading.Activities.Services/Extensions.cs (hunk 9 above)
Comment (andrey-anokhin) [reply src/MarginTrading.Activities.Services/Extensions.cs]: In the real world asset pair accuracy cannot be less than zero for sure ) But this method accepts only int value. Who knows what we will receive ) And it is not a big difference to use > or = ).

File: src/MarginTrading.Activities.Services/Projections/OrdersProjection.cs (hunk 11 above)
Comment (andrey-anokhin) [reply src/MarginTrading.Activities.Services/Projections/OrdersProjection.cs]: No difference. Both of them are async internally and obsolete)

File: src/MarginTrading.Activities.Services/Projections/PositionsProjection.cs (hunk 12 above)
Comment (andrey-anokhin) [reply src/MarginTrading.Activities.Services/Projections/PositionsProjection.cs]: No difference. Both of them are async internally and obsolete)

File: src/MarginTrading.Activities.SqlRepositories/ActivitiesRepository.cs (hunk 13 above)
Comment (andrey-anokhin) [reply src/MarginTrading.Activities.SqlRepositories/ActivitiesRepository.cs]: No difference too )

File: src/MarginTrading.Activities.Services/Projections/OrdersProjection.cs (hunk 11 above)
Comment (GSailen) [reply src/MarginTrading.Activities.Services/Projections/OrdersProjection.cs]: That's a horror

**Pull request merged**